harvester/parse.py

```diff
@@ -0,0 +1,442 @@
+        the WARC file and the offset (where the binary website recording begins), but this
+        library takes it from there, understanding where in the WARC file the record ends,
+        and providing a handy reader of the actual data.
+
+        In the event we ever want to explore extracting other file types (e.g. media,
+        JSON, etc.) this approach will also prove useful.
+        """
+        with self._get_archive_file_obj(warc_filepath) as file_obj:
+            file_obj.seek(offset)
+            record = next(ArchiveIterator(file_obj))
+            yield record
+
+    # ruff: noqa: FBT001, FBT002
+    def get_website_content_by_url(self, url: str, decode: bool = True) -> str:
+        """Get HTML content via a URL.
+
+        This is slower than using self.get_website_content() if the warc filename and
+        offset are known, as this needs to search the dataframe for the URL to get the
+        WARC filename.
+        """
+        # get warc file the URL belongs to
+        rows = self.websites_df[self.websites_df.url == url]
+        if len(rows) == 1:
+            row = rows.iloc[0]
+        else:
+            exc_msg = f"Could not find url in CDX index: {url}"
+            raise FileNotFoundError(exc_msg)
+        return self.get_website_content(row.filename, row.offset, decode=decode)
+
+    def get_website_content(
+        self, warc_filename: str, offset: str | int, decode: bool = True
+    ) -> str:
+        """Extract HTML content from a WARC record."""
+        warc_filepath = f"{self.WARC_DIR}/{warc_filename}"
+        with self._get_warc_record(warc_filepath, int(offset)) as record:
+            content = record.content_stream().read()
+            if decode:
+                content = content.decode("utf-8")
+            return content
+
+    @classmethod
+    def get_html_content_metadata(cls, html_content: str | bytes) -> dict:
+        """Extract structured metadata from the website HTML."""
+        soup = BeautifulSoup(html_content, "html.parser")
+        tags = {}
+        for og_tag_name in cls.HTML_CONTENT_METADATA_TAGS:
+            og_tag = soup.find("meta", property=og_tag_name)
+            if og_tag is None:
+                continue
+            if content := og_tag.get("content"):  # type: ignore[union-attr]
+                content_stripped = content.strip()  # type: ignore[union-attr]
+                if content_stripped != "":
+                    og_tag_name_friendly = og_tag_name.replace(":", "_")
+                    tags[og_tag_name_friendly] = content_stripped
+        return tags
+
+    def _parse_fulltext(self, fulltext: str) -> str:
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
+        return fulltext
+
+    def _parse_fulltext_keywords(self, fulltext: str) -> str:
```

Comment:
Same here about `parse`, I feel like it's covering different actions/concepts here which can muddle things even if it's technically accurate, what about `extract`?